### PR TITLE
LidoOracle

### PR DESCRIPTION
**Pull request by alcueca**
No tests yet, but good as a first stab to agree on coding practices.

Once we get a couple of these done, they will come out quickly.

### CODE REVIEW

**Review by ruvaag — COMMENTED** [review]
no errors.

some nits:
- a short comment on if-else branches can help first-time readers internalise terms like base, quote, scalar, etc. more easily.  [example from erc4626 oracle](https://github.com/ruvaag/oracles/blob/21c9cb0bd950dbf5f55a3236b92d50d10107f452/src/erc4626/ERC4626Oracle.sol#L36)
- we don't need BoringERC20 for this particular case. it's verifiable that `.decimals()` will not revert.

repo-wide practices:
- a way to organise interfaces: interfaces that are used by more than one oracle should be stored in `src/interfaces` and those related to a particular oracle must go in the oracle directory. this PR already follows the pattern  -`IERC20` goes in the common `interfaces` while IWSTETH is in `src/lido`. any thoughts?

**Comment by totomanov** [comment]
It would be cheaper to just call the corresponding conversion methods on the stETH contract.

See WstETH [source](https://etherscan.io/address/0x7f39c581f595b53c5cb19bd0b3f8da6c935e2ca0#code#L1111):
```solidity
    /**
     * @notice Get amount of wstETH for a given amount of stETH
     * @param _stETHAmount amount of stETH
     * @return Amount of wstETH for a given stETH amount
     */
    function getWstETHByStETH(uint256 _stETHAmount) external view returns (uint256) {
        return stETH.getSharesByPooledEth(_stETHAmount);
    }

    /**
     * @notice Get amount of stETH for a given amount of wstETH
     * @param _wstETHAmount amount of wstETH
     * @return Amount of stETH for a given wstETH amount
     */
    function getStETHByWstETH(uint256 _wstETHAmount) external view returns (uint256) {
        return stETH.getPooledEthByShares(_wstETHAmount);
    }
```

File: src/lido/LidoOracle.sol [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {IERC20} from "../interfaces/IERC20.sol";
+import {IWSTETH} from "./IWSTETH.sol";
+import {BoringERC20} from "../libraries/BoringERC20.sol";
+
+contract LidoOracle {
+    using BoringERC20 for IERC20;
+    error OracleUnsupportedPair();
```
Comment (totomanov) [inline src/lido/LidoOracle.sol:10]: To abide by the spec, this error should be `error OracleUnsupportedPair(address base, address quote);`

File: src/libraries/BoringERC20.sol [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: MIT
+// With thanks to @BoringCrypto
+pragma solidity ^0.8.0;
+import "../interfaces/IERC20.sol";
```
Comment (totomanov) [inline src/libraries/BoringERC20.sol:4]: We can use the IERC20 interface in forge-std. This way the local one can be removed.

`import "forge-std/interfaces/IERC20.sol";`

File: src/lido/LidoOracle.sol [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {IERC20} from "../interfaces/IERC20.sol";
```
Comment (totomanov) [inline src/lido/LidoOracle.sol:4]: We can use the IERC20 interface in forge-std. This way the local one can be removed.

import "forge-std/interfaces/IERC20.sol";

**Comment by alcueca** [comment]
I will recreate this code from `main` once #1 is merged, taking into account the excellent comments here.

**Pull request closed without merging**